type-metadata/src/lib.rs

```diff
@@ -1,7 +1,346 @@
-#[cfg(test)]
-mod tests {
-	#[test]
-	fn it_works() {
-		assert_eq!(2 + 2, 4);
+// Copyright 2019 Centrality Investments Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod registry;
+pub use registry::Registry;
+
+mod tests;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum IdentKind {
+	Custom(CustomIdent),
+
+	// primitives, including build-in types and common precludes
+	Bool,
+	Str,
+	U8,
+	U16,
+	U32,
+	U64,
+	U128,
+	I8,
+	I16,
+	I32,
+	I64,
+	I128,
```

Comment:
I would simplify `IdentKind` further by putting all primitives into a variant like this:

```rust
pub enum IdentKind {
    ...
    Primitive(PrimitiveIdent),
    ...
}

pub enum PrimitiveIdent {
	Bool,
	Str,
	U8,
	U16,
	U32,
	U64,
	U128,
	I8,
	I16,
	I32,
	I64,
	I128,
}
```

And then do a `From<PrimitiveIdent>` impl for `IdentKind`. For this you can use `derive_more` crate to reduce boilerplate code to a minimum.

This design helps in dispatching for primitive types so we can simply pattern match them against their main case.
Also if we encode this structure correctly we should have no overhead either.

Reply:
It does make no overhead binary encoding harder.
With current approach, 1 byte is enough for primitive types. If a new type is used, the naive encoding format will require 2 bytes. To make it fit in 1 byte, I guess we can have 3 bits for `IdentKind` and 5 bits for `PrimitiveIdent`?

Reply:
@shaopengw you need to make sure the primitives are at end so adding more primitives at end is making more sense.

Reply:
Why was this marked resolved? I thought we just had to implement the en-/decoding for this with the proposal of XLC so resolve the increase in binary size which should be easy. I would really prefer having a better structured kinds enum that is easier to pattern match in some cases.